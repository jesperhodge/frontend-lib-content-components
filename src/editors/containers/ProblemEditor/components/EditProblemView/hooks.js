import { useState } from 'react';
import 'tinymce';
import { StrictDict } from '../../../../utils';
import ReactStateSettingsParser from '../../data/ReactStateSettingsParser';
import ReactStateOLXParser from '../../data/ReactStateOLXParser';
import { setAssetToStaticUrl } from '../../../../sharedComponents/TinyMceWidget/hooks';
import { ProblemTypeKeys } from '../../../../data/constants/problem';

export const state = StrictDict({
  isNoAnswerModalOpen: (val) => useState(val),
});

export const noAnswerModalToggle = () => {
  const [isNoAnswerModalOpen, setIsOpen] = state.isNoAnswerModalOpen(false);
  return {
    isNoAnswerModalOpen,
    openNoAnswerModal: () => setIsOpen(true),
    closeNoAnswerModal: () => setIsOpen(false),
  };
};

export const fetchEditorContent = ({ format }) => {
  const editorObject = { hints: [] };
  const EditorsArray = window.tinymce.editors;
  Object.entries(EditorsArray).forEach(([id, editor]) => {
    if (Number.isNaN(parseInt(id))) {
      if (id.startsWith('answer')) {
        const { answers } = editorObject;
        const answerId = id.substring(id.indexOf('-') + 1);
        editorObject.answers = { ...answers, [answerId]: editor.getContent({ format }) };
      } else if (id.includes('Feedback')) {
        const { selectedFeedback, unselectedFeedback, groupFeedback } = editorObject;
        const feedbackId = id.substring(id.indexOf('-') + 1);
        if (id.startsWith('selected')) {
          editorObject.selectedFeedback = { ...selectedFeedback, [feedbackId]: editor.getContent({ format }) };
        }
        if (id.startsWith('unselected')) {
          editorObject.unselectedFeedback = { ...unselectedFeedback, [feedbackId]: editor.getContent({ format }) };
        }
        if (id.startsWith('group')) {
          editorObject.groupFeedback = { ...groupFeedback, [feedbackId]: editor.getContent({ format }) };
        }
      } else if (id.startsWith('hint')) {
        const { hints } = editorObject;
        hints.push(editor.getContent({ format }));
      } else {
        editorObject[id] = editor.getContent();
      }
    }
  });
  return editorObject;
};

export const parseState = ({
  problem,
  isAdvanced,
  ref,
  assets,
  lmsEndpointUrl,
}) => () => {
  const editorObject = fetchEditorContent({ format: '' });
  const reactSettingsParser = new ReactStateSettingsParser(problem);
  const reactOLXParser = new ReactStateOLXParser({ problem, editorObject });
  const reactBuiltOlx = setAssetToStaticUrl({ editorValue: reactOLXParser.buildOLX(), assets, lmsEndpointUrl });
  const rawOLX = ref?.current?.state.doc.toString();
  return {
    settings: reactSettingsParser.getSettings(),
    olx: isAdvanced ? rawOLX : reactBuiltOlx,
  };
};

export const checkForNoAnswers = ({ openNoAnswerModal, problem }) => {
  const simpleTextAreaProblems = [ProblemTypeKeys.DROPDOWN, ProblemTypeKeys.NUMERIC, ProblemTypeKeys.TEXTINPUT];
  const editorObject = fetchEditorContent({ format: '' });
  const { problemType } = problem;
  const { answers } = problem;
  const answerTitles = simpleTextAreaProblems.includes(problemType) ? {} : editorObject.answers;

  const hasTitle = () => {
    const titles = [];
    answers.forEach(answer => {
      const title = simpleTextAreaProblems.includes(problemType) ? answer.title : answerTitles[answer.id];
      if (title.length > 0) {
        titles.push(title);
      }
    });
    if (titles.length > 0) {
      return true;
    }
    return false;
  };

  const hasNoCorrectAnswer = () => {
    let correctAnswer;
    answers.forEach(answer => {
      if (answer.correct) {
        const title = simpleTextAreaProblems.includes(problemType) ? answer.title : answerTitles[answer.id];
        if (title.length > 0) {
          correctAnswer = true;
        }
      }
    });
    if (correctAnswer) {
      return true;
    }
    return false;
  };

  if (problemType === ProblemTypeKeys.NUMERIC && !hasTitle()) {
    openNoAnswerModal();
    return true;
  }
  if (!hasNoCorrectAnswer()) {
    openNoAnswerModal();
    return true;
  }
  return false;
};

export const getContent = ({
  problemState,
  openNoAnswerModal,
  isAdvancedProblemType,
  editorRef,
  assets,
  lmsEndpointUrl,
}) => {
  const problem = problemState;
  const hasNoAnswers = isAdvancedProblemType ? false : checkForNoAnswers({
    problem,
    openNoAnswerModal,
  });
  if (!hasNoAnswers) {
    const data = parseState({
      isAdvanced: isAdvancedProblemType,
      ref: editorRef,
      problem,
      assets,
      lmsEndpointUrl,
    })();
    return data;
  }
  return null;
};
