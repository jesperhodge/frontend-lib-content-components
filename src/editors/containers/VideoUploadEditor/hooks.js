import * as requests from '../../data/redux/thunkActions/requests';

export const uploadVideo = async ({ dispatch, supportedFiles }) => {
  const data = { files: [] };
  supportedFiles.forEach((file) => {
    data.files.push({
      file_name: file.name,
      content_type: file.type,
    });
  });
  dispatch(await requests.uploadVideo({
    data,
    onSuccess: async (response) => {
      const { files } = response.json();
      await Promise.all(Object.values(files).map(async (fileObj) => {
        const fileName = fileObj.file_name;
        const uploadUrl = fileObj.upload_url;
        const uploadFile = supportedFiles.find((file) => file.name === fileName);
        if (!uploadFile) {
          console.error(`Could not find file object with name "${fileName}" in supportedFiles array.`);
          return;
        }
        const formData = new FormData();
        formData.append('uploaded-file', uploadFile);
        await fetch(uploadUrl, {
          method: 'PUT',
          body: formData,
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        })
          .then((resp) => resp.json())
          .then((responseData) => console.log('File uploaded:', responseData))
          .catch((error) => console.error('Error uploading file:', error));
      }));
    },
  }));
};

export default {
  uploadVideo,
};
