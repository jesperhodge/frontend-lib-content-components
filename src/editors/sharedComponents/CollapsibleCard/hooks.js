import { useState } from 'react';

/* eslint-disable import/prefer-default-export */
export const useCollapsible = () => {
  const [isCardCollapsibleOpen, setIsCardCollapsibleOpen] = useState(false);
  return {
    isCardCollapsibleOpen,
    toggleCardCollapse: () => setIsCardCollapsibleOpen(!isCardCollapsibleOpen),
  };
};
