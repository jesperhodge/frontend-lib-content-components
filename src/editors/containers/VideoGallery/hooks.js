import React from 'react';
import * as module from './hooks';
import messages from './messages';
import {
  filterKeys,
  filterMessages,
  sortKeys,
  sortMessages,
  sortFunctions,
} from './utils';

export const state = {
  highlighted: (val) => React.useState(val),
  searchString: (val) => React.useState(val),
  showSelectVideoError: (val) => React.useState(val),
  showSizeError: (val) => React.useState(val),
  sortBy: (val) => React.useState(val),
  filertBy: (val) => React.useState(val),
  hideSelectedVideos: (val) => React.useState(val),
};

export const searchAndSortProps = () => {
  const [searchString, setSearchString] = module.state.searchString('');
  const [sortBy, setSortBy] = module.state.sortBy(sortKeys.dateNewest);
  const [filterBy, setFilterBy] = module.state.filertBy(filterKeys.videoStatus);
  const [hideSelectedVideos, setHideSelectedVideos] = module.state.hideSelectedVideos(false);

  return {
    searchString,
    onSearchChange: (e) => setSearchString(e.target.value),
    clearSearchString: () => setSearchString(''),
    sortBy,
    onSortClick: (key) => () => setSortBy(key),
    sortKeys,
    sortMessages,
    filterBy,
    onFilterClick: (key) => () => setFilterBy(key),
    filterKeys,
    filterMessages,
    showSwitch: true,
    hideSelectedVideos,
    switchMessage: messages.hideSelectedCourseVideosSwitchLabel,
    onSwitchClick: () => setHideSelectedVideos(!hideSelectedVideos),
  };
};

export const filterListBySearch = ({ searchString, videoList }) => (
  videoList.filter(({ displayName }) => displayName.toLowerCase().includes(searchString.toLowerCase()))
);

export const filterListByStatus = ({ statusFilter, videoList }) => {
  if (statusFilter === filterKeys.videoStatus) {
    return videoList;
  }
  return videoList.filter(({ status }) => status === statusFilter);
};

export const filterListByHideSelectedCourse = ({ videoList }) => (
  // TODO Missing to implement this
  videoList
);

export const filterList = ({
  sortBy,
  filterBy,
  searchString,
  videos,
}) => {
  let filteredList = module.filterListBySearch({
    searchString,
    videoList: videos,
  });
  filteredList = module.filterListByStatus({
    statusFilter: filterBy,
    videoList: filteredList,
  });
  filteredList = module.filterListByHideSelectedCourse({
    videoList: filteredList,
  });
  return filteredList.sort(sortFunctions[sortBy in sortKeys ? sortKeys[sortBy] : sortKeys.dateNewest]);
};

export const videoListProps = ({ searchSortProps, videos }) => {
  const [highlighted, setHighlighted] = module.state.highlighted(null);
  const [
    showSelectVideoError,
    setShowSelectVideoError,
  ] = module.state.showSelectVideoError(false);
  const [
    showSizeError,
    setShowSizeError,
  ] = module.state.showSizeError(false);
  const filteredList = module.filterList({ ...searchSortProps, videos });
  return {
    galleryError: {
      show: showSelectVideoError,
      set: () => setShowSelectVideoError(true),
      dismiss: () => setShowSelectVideoError(false),
      message: messages.selectVideoError,
    },
    // TODO We need to update this message when implementing the video upload screen
    inputError: {
      show: showSizeError,
      set: () => setShowSizeError(true),
      dismiss: () => setShowSelectVideoError(false),
      message: messages.fileSizeError,
    },
    galleryProps: {
      galleryIsEmpty: Object.keys(filteredList).length === 0,
      searchIsEmpty: filteredList.length === 0,
      displayList: filteredList,
      highlighted,
      onHighlightChange: (e) => setHighlighted(e.target.value),
      emptyGalleryLabel: messages.emptyGalleryLabel,
      showIdsOnCards: true,
      height: '100%',
    },
    selectBtnProps: {
      onclick: () => {
        // TODO Update this when implementing the selection feature
      },
    },
  };
};

export const fileInputProps = () => {
  // TODO [Update video] Implement this
  const ref = React.useRef();
  const click = () => ref.current.click();

  return {
    click,
    addFile: () => {},
    ref,
  };
};

export const buildVideos = ({ rawVideos }) => {
  let videos = [];
  const rawVideoList = Object.values(rawVideos);
  if (rawVideoList.length > 0) {
    videos = rawVideoList.map(video => ({
      id: video.edx_video_id,
      displayName: video.client_video_id,
      externalUrl: video.course_video_image_url,
      dateAdded: video.created,
      locked: false,
      thumbnail: video.course_video_image_url,
      status: video.status,
      statusBadgeVariant: module.getstatusBadgeVariant({ status: video.status }),
      duration: video.duration,
      transcripts: video.transcripts,
    }));
  }
  return videos;
};

export const getstatusBadgeVariant = ({ status }) => {
  switch (status) {
    case filterKeys.failed:
      return 'danger';
    case filterKeys.uploading:
    case filterKeys.processing:
      return 'light';
    default:
      return null;
  }
};

export const videoProps = ({ videos }) => {
  const searchSortProps = module.searchAndSortProps();
  const videoList = module.videoListProps({ searchSortProps, videos });
  const {
    galleryError,
    galleryProps,
    inputError,
    selectBtnProps,
  } = videoList;
  const fileInput = module.fileInputProps();

  return {
    galleryError,
    inputError,
    fileInput,
    galleryProps,
    searchSortProps,
    selectBtnProps,
  };
};

export default {
  videoProps,
  buildVideos,
};
