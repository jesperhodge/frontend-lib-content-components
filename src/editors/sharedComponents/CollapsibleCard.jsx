import React from 'react';
import { Collapsible, Icon, Card } from '@edx/paragon';
import { KeyboardArrowUp, KeyboardArrowDown } from '@edx/paragon/icons';
import {
  arrayOf, string, node, bool,
} from 'prop-types';
import { useFullCard } from '../containers/ProblemEditor/components/EditProblemView/SettingsWidget/hooks';

const CardSection = ({
  children, none, isCardCollapsibleOpen, summary,
}) => {
  const show = isCardCollapsibleOpen || summary;
  if (!show) { return null; }

  return (
    <Card.Section className="px-4 pb-4 pt-3">
      <Collapsible.Advanced
        open={!isCardCollapsibleOpen}
      >
        <Collapsible.Body className="collapsible-body">
          <span className={`small ${none ? 'text-gray-500' : 'text-primary-500'}`}>{summary}</span>
        </Collapsible.Body>
      </Collapsible.Advanced>
      <Collapsible.Advanced
        open={isCardCollapsibleOpen}
      >
        <Collapsible.Body className="collapsible-body">
          {children}
        </Collapsible.Body>
      </Collapsible.Advanced>
    </Card.Section>
  );
};
CardSection.propTypes = {
  none: bool,
  children: node.isRequired,
  summary: string,
  isCardCollapsibleOpen: bool.isRequired,
};
CardSection.defaultProps = {
  none: false,
  summary: null,
};

/** CollapsibleCard
 *
 * You can pass an array of sections to this component, in place of children.
 * The elements of the sections array will each be rendered in their own CardSection.
 */
export const CollapsibleCard = ({
  title, className, children, summary, sections, ...passThroughProps
}) => {
  const { isCardCollapsibleOpen, toggleCardCollapse } = useFullCard();
  const content = children ? [children, ...sections] : sections;

  return (
    <Card className={`${className} CollapsibleCard border border-light-700 shadow-none`}>
      <Card.Section className="settingsCardTitleSection">
        <Collapsible.Advanced
          open={isCardCollapsibleOpen}
          onToggle={toggleCardCollapse}
        >
          <Collapsible.Trigger className="collapsible-trigger d-flex">
            <span className="flex-grow-1 text-primary-500 x-small">{title}</span>
            <Collapsible.Visible whenClosed>
              <Icon src={KeyboardArrowDown} />
            </Collapsible.Visible>
            <Collapsible.Visible whenOpen>
              <Icon src={KeyboardArrowUp} />
            </Collapsible.Visible>
          </Collapsible.Trigger>
        </Collapsible.Advanced>
      </Card.Section>
      {content.map((section, index) => (
        <>
          {isCardCollapsibleOpen && <hr />}
          {/* eslint-disable-next-line react/no-array-index-key */}
          <CardSection {...passThroughProps} isCardCollapsibleOpen={isCardCollapsibleOpen} key={`CollapsibleCard-${title}-${index}`}>
            {section.children}
          </CardSection>
        </>
      ))}
    </Card>
  );
};
CollapsibleCard.propTypes = {
  title: string.isRequired,
  children: node,
  sections: arrayOf(node),
  className: string,
  summary: string.isRequired,
};
CollapsibleCard.defaultProps = {
  className: '',
  children: null,
  sections: [],
};

export default CollapsibleCard;
