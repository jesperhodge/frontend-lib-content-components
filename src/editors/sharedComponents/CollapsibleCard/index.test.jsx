import React from 'react';
import { shallow } from 'enzyme';
import CollapsibleCard from './index';
import { render, screen, findByText } from '@testing-library/react';
import { useCollapsible } from './hooks';
import '@testing-library/jest-dom';

const MockComponent = () => {
  return <div><p>Text</p></div>;
}

const Card = () => <MockComponent />;
Card.Section = () => <MockComponent />;

jest.mock('./CardSection', () => () => <MockComponent />);
jest.mock('@edx/paragon', () => ({
  __esModule: true,
  Collapsible: {
    Advanced: () => <MockComponent />,
    Visible: () => <MockComponent />,
    Trigger: () => <MockComponent />,
  },
  Icon: () => <MockComponent />,
  Card,
}));
jest.mock('@edx/paragon/icons', () => ({
  __esModule: true,
  KeyboardArrowUp: () => <MockComponent />,
  KeyboardArrowDown: () => <MockComponent />,
}));

describe('CollapsibleCard', () => {
  describe('default with children', () => {
    const children = (<h1>My test content</h1>);
    test('snapshot: renders correct', () => {
      expect(shallow(<CollapsibleCard title="Settings Option Title" summary="Settings Option Summary">{children}</CollapsibleCard>)).toMatchSnapshot();
    });
    test('opens', () => {
      render(<CollapsibleCard title="Settings Option Title" summary="Settings Option Summary">{children}</CollapsibleCard>);
      expect(screen.getByText('Text')).toBeInTheDocument();
    });
    test('MockComponent', () => {
      render(<MockComponent />);
      expect(screen.getByText('Text')).toBe('Text');
    });
  });
  describe('with additional sections', () => {
    const children = (<h1>First Section</h1>);
    const sections = [<h1>Second Section</h1>, <h1>Third Section</h1>];
    test('snapshot: renders correct', () => {
      expect(shallow(
        <CollapsibleCard title="Settings Option Title" summary="Settings Option Summary" extraSections={sections}>
          {children}
        </CollapsibleCard>,
      )).toMatchSnapshot();
    });
  });
});
