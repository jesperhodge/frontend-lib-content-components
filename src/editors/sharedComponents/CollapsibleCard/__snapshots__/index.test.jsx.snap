// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleCard default with children snapshot: renders correct 1`] = `
<Card
  className=" CollapsibleCard border border-light-700 shadow-none"
>
  <Card.Section
    className="settingsCardTitleSection"
  >
    <Advanced
      onToggle={[Function]}
      open={false}
    >
      <Trigger
        className="collapsible-trigger d-flex"
      >
        <span
          className="flex-grow-1 text-primary-500 x-small"
        >
          Settings Option Title
        </span>
        <Visible
          whenClosed={true}
        >
          <Icon />
        </Visible>
        <Visible
          whenOpen={true}
        >
          <Icon />
        </Visible>
      </Trigger>
    </Advanced>
  </Card.Section>
  <CardSection
    isCardCollapsibleOpen={false}
    key="CollapsibleCard-Settings Option Title-0"
    none={false}
    summary={null}
  />
</Card>
`;

exports[`CollapsibleCard with additional sections snapshot: renders correct 1`] = `
<Card
  className=" CollapsibleCard border border-light-700 shadow-none"
>
  <Card.Section
    className="settingsCardTitleSection"
  >
    <Advanced
      onToggle={[Function]}
      open={false}
    >
      <Trigger
        className="collapsible-trigger d-flex"
      >
        <span
          className="flex-grow-1 text-primary-500 x-small"
        >
          Settings Option Title
        </span>
        <Visible
          whenClosed={true}
        >
          <Icon />
        </Visible>
        <Visible
          whenOpen={true}
        >
          <Icon />
        </Visible>
      </Trigger>
    </Advanced>
  </Card.Section>
  <CardSection
    extraSections={
      Array [
        <h1>
          Second Section
        </h1>,
        <h1>
          Third Section
        </h1>,
      ]
    }
    isCardCollapsibleOpen={false}
    key="CollapsibleCard-Settings Option Title-0"
    none={false}
    summary={null}
  />
</Card>
`;
